torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -478,6 +494,108 @@ def _low_precision_hook_enabled(state: _State) -> bool:
     return state._communication_hook in LOW_PRECISION_HOOKS
 
 
+@no_type_check
+@torch.no_grad()
+def _post_backward_callback(
+    state: _State,
+):
+    """
+    This waits for the post-backward to finish and performs some final cleanup.
+    This should only be called on the root FSDP instance.
+    """
+    p_assert(
+        state._is_root,
+        "The post-backward callback should only be called on the root FSDP instance",
+    )
+
+    if state._sync_gradients:
+        torch.cuda.current_stream().wait_stream(state._streams["post_backward"])
+        if state.cpu_offload.offload_params:
+            # Wait for non-blocking GPU -> CPU sharded gradient copies from the
+            # post-backward hooks to finish explicitly since CPU gradients do
+            # not automatically synchronize with the GPU
+            torch.cuda.current_stream().synchronize()
+    state._exec_order_data.next_iter()
+
+    states = [state] if _is_composable(state) else state.fsdp_modules(state)
+    for state in states:
+        _catch_all_reshard(state)
+        _finalize_params(state)
+        state._ran_pre_backward_hook.clear()
+        state.training_state = TrainingState.IDLE
+        for handle in state._handles:
+            handle._training_state = HandleTrainingState.IDLE
+        state._handles_prefetched.clear()
+    # Reset for cases like one forward and multiple backwards
+    state._post_backward_callback_queued = False
+
+
```

Comment:
I pulled these methods out from being only defined inside the post-backward callback. It does not matter either way, but it might be easier to read when everything is not so indented from being nested.